fix(selection-list): fix option value coercion and selection events

dereekb opened the pull request:
* md-list-option value is no longer coerced to a boolean value
* md-list-option EventEmitters selectedChange and deselected now emit an event when an option is selected/deselected

Fixes #6864

willshowell commented on src/lib/list/selection-list.ts:
```diff
@@ -97,7 +97,7 @@ export class MdListOption extends _MdListOptionMixinBase
   @Input()
   get value() { return this._value; }
-  set value( val: any) { this._value = coerceBooleanProperty(val); }
+  set value( val: any) { this._value = val; }
```
The getter and setter can be removed if you're no longer coercing

rafaelss95 commented:
I just want to ask something: why is there an event for `select` and also for `deselect`?

I think it can be just one event, like in checkbox:

  ```
private _emitChangeEvent() {
    let event = new MdCheckboxChange();
    event.source = this;
    event.checked = this.checked;

    this._controlValueAccessorChangeFn(this.checked);
    this.change.emit(event);
  }
```

Also, is the `(destroyed)` useful at some point?

jelbourn commented:
Yeah, that's a good point. It should just have one `selectionChanged` event where the event includes the new state.

dereekb commented:
In my current project I am watching the destroyed event so I can remove options that are no longer available. You could probably use the changes QueryList in selectionList, but you'd have to figure out which ones were removed.

Looking at the Material code again, the MdSelectionList itself subscribes to the destroyed events.

rafaelss95 commented:
@jelbourn what do you think about other statements?

jelbourn commented:
@rafaelss95 those should be covered under #6677

dereekb commented:
I went ahead and pushed changes that were relevant to the original issue.

I didn't rename MdSelectionListOptionEvent, but I changed it to a class now so it stayed consistent with similar classes.

ptretyakov commented:
Hello, Everybody!
When are you merge the pull request? 
Thank you for your work :)

mmalerba commented:
please rebase

dereekb commented:
rebase with master or _presubmit? Looks like someone else made some of the the same changes

dereekb commented:
Scratch that, I broke the fixed

dereekb commented:
Ok, rebased with changes #6983 and #6971 included.

kara commented:
@dereekb Would you mind rebasing one more time?

dereekb commented:
Sure.

I'm rebasing right now, but it looks like a lot of the changes in this pull request are already on master one way or another from other pull requests.

The only changes are changing MatSelectionListOptionEvent to a class, and merging selectChange and deselected together into a single event emitter, that last one being a breaking change.

dereekb commented:
Noticed I need to rebase it again. Working on that now.

dereekb commented:
Ok, rebased and all the unit tests passed on my end.

I want to bring up a slight change from this rebase:

I removed _selected and instead have the getter use the selectionList's selectedOptions isSelected() for whether or not the object is selected.

My reasoning is that the selection's state shouldn't be in two places (by this I mean MatSelectionList and other components check selectedOptions for is selected, while the MatListOption checks itself), and the performance change is minimal since a Set is used for selectedOptions.

---

Another thing, since selectedOptions in MatSelectionList is publicly accessible, externally modifying selectedOptions should also send selectionChange events on the MatListOption right? I don't think it would, so it's something to consider.

adarowiec commented:
Can you guys already pull it? I really need this fix in my project.

rafaelss95 commented on src/lib/list/selection-list.spec.ts:
```diff
@@ -579,3 +658,28 @@ class SelectionListWithTabindexBinding {
 }
+
+@Component({template: `
+<mat-selection-list id = "selection-list-5">
```
remove the around spaces: `id="selection-list-5"`?

dereekb replied:
I don't know if I hit the format button on my side, but there are spaces around most of the unit test component's id's in this file.

I'll fix them.

rafaelss95 replied:
You're right. There are a lot of strange spaces in this spec file.

rafaelss95 commented on src/lib/list/selection-list.ts:
```diff
@@ -174,6 +169,16 @@ export class MatListOption extends _MatListOptionMixinBase
 
+  /** Creates a selection event object from the specified option. */
+  private _createChangeEvent(option: MatListOption = this): MatSelectionListOptionEvent {
+    let event = new MatSelectionListOptionEvent();
```
`const`?

rafaelss95 commented on src/lib/list/selection-list.ts:
```diff
@@ -52,8 +52,9 @@ export class MatListOptionBase {}
+export class MatSelectionListOptionEvent {
   option: MatListOption;
+  selected: boolean;
 }
```
I'd keep the consistency between this and others:

/** Change event object emitted by MatSelectionList. */
export class MatSelectionListChange {
  /** The source MatListOption of the event. */
  source: MatListOption;
  /** The new `checked` value of the option. */
  checked: boolean;
}

See [https://github.com/angular/material2/blob/master/src/lib/checkbox/checkbox.ts#L71](checkbox for example).

dereekb replied:
So checked instead of selected? Gotcha.

rafaelss95 replied:
Also adding comments will be useful so it can be shown in docs.

dereekb replied:
By others you mean other checkboxes? Add comments?

Changing it from selected to checked seems like a step in the wrong direction, since it is MatListOption.selected, not MatListOption.checked.

rafaelss95 replied:
You're right. But what about `source` instead of `option`? All `change` events use `source`.

dereekb replied:
Sounds good, I'll change that.

rafaelss95 commented on src/lib/list/selection-list.ts:
```diff
@@ -97,34 +97,29 @@ export class MatListOption extends _MatListOptionMixinBase
   /** Whether the option is selected. */
   @Input()
-  get selected() { return this._selected; }
+  get selected() { return this.selectionList.selectedOptions.isSelected(this); }
```
Can you add `:boolean` here?

rafaelss95 commented on src/lib/list/selection-list.ts:
```diff
@@ -97,34 +97,29 @@ export class MatListOption extends _MatListOptionMixinBase
-
   /** Whether the option is disabled. */
   @Input()
   get disabled() { return (this.selectionList && this.selectionList.disabled) || this._disabled; }
```
`: boolean`?

rafaelss95 commented on src/lib/list/selection-list.ts:
```diff
@@ -97,34 +97,29 @@ export class MatListOption extends _MatListOptionMixinBase
   /** Whether the option is disabled. */
   @Input()
   get disabled() { return (this.selectionList && this.selectionList.disabled) || this._disabled; }
   set disabled(value: any) { this._disabled = coerceBooleanProperty(value); }
```
`: boolean`?

dereekb replied:
Value is the actual selection option value. 

We don't want to coerce it to a boolean, again! :stuck_out_tongue:

rafaelss95 replied:
I'm not talking about `value`, but `disabled`.

rafaelss95 replied:
In `selected` we have: `set selected(value: boolean) {`

dereekb replied:
Gotcha, fixed.

rafaelss95 replied:
I think you misunderstood what I mean... I mean that we have the following signature for selected: `set selected(value: boolean) { ... = coerceBooleanProperty(value); }`, so the `disabled` should follow the same thing: `set disabled(value: boolean) { ... = coerceBooleanProperty(value); }`. Did you get it now?

dereekb replied:
Ah, ok. The highlighted code I see is both on this section:

```
/** Value of the option */
-  @Input() value: any;
```

So I was kind of guessing what you meant. Making those changes now.

rafaelss95 commented on src/lib/list/selection-list.ts:
```diff
@@ -111,7 +111,7 @@ export class MatListOption extends _MatListOptionMixinBase
   get selected(): boolean { return this.selectionList.selectedOptions.isSelected(this); }
   set selected(value: boolean) {
-    const isSelected = coerceBooleanProperty(value);
+    const isSelected: boolean = coerceBooleanProperty(value);
```
I think you misunderstood what I mean... I mean that we have the following signature for selected: `set selected(value: boolean) { ... = coerceBooleanProperty(value); }`, so the `disabled` should follow the same thing: `set disabled(value: boolean) { ... = coerceBooleanProperty(value); }`. 

Basically change the signature of disabled from `set disabled(value: any) {` to `set disabled(value: boolean) {`.

 Did you get it now?

dereekb replied:
Yea, the highlighted section GitHub was showing was:

```
/** Value of the option */
-  @Input() value: any;
```

So I didn't really get what you were talking about. It's changes now and I removed the unnecessary :boolean.

rafaelss95 replied:
Aha, ok. Nice :)

rafaelss95 replied:
Aha, ok. Nice :)

Will you remove this `:boolean` in another commit?

dereekb replied:
That's the second time today I've committed without first hitting the save button. Done.

rafaelss95 commented on src/lib/list/selection-list.ts:
```diff
@@ -51,16 +51,18 @@ export const _MatSelectionListMixinBase =
 
-/** Event emitted by a selection-list whenever the state of an option is changed. */
+/** Change event object emitted by MatListOption */
 export class MatSelectionListOptionEvent {
```
`MatListOptionChange`?

andrewseguin commented:
Can you check on the tests and get them green?

dereekb commented:
The tests that failed aren't related to the changes (they have to do with MatAutocomplete).

Those failed tests are related to #7743.

alisonmsmith commented:
FYI - material docs specify the event output as 'selectChange' which is of type EventEmitter<MatSelectionListOptionEvent> and instead it looks like it is implemented here as 'selectionChange' of type EventEmitter<MatListOptionChange> also the 'deselected' event from the docs is still not emitted (although you can check if the option was selected or deselected from the MatListOptionChange boolean selected property)

dereekb commented:
Yea, the docs need to be updated to reflect the changes.

selected/deselected were removed and replaced with selectionChange, since having a single selectionChange is more consistent in terms of the rest of Angular Material implementations.

As for the selectChange/selectionChange name, there was another discussion thread where I believe selectionChange was picked as the more desirable. I don't see the reference in this thread anymore, but it was about naming consistency and updating across a number of different Material components.
